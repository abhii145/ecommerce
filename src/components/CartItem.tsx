import { FaMinus, FaPlus, FaTrashAlt } from "react-icons/fa"
import { CartItemProps } from "../types"

const CartItem: React.FC<CartItemProps> = ({
  item,
  onIncrease,
  onDecrease,
  onRemove,
}) => {
  return (
    <div className="flex flex-col md:flex-row items-center justify-between border-b py-4">
      <div className="flex items-center space-x-4 mb-4 md:mb-0">
        <img
          src={item.thumbnail}
          alt={item.title}
          className="w-24 h-24 object-cover rounded-md"
          loading="lazy"
        />
        <div>
          <h3 className="text-lg font-semibold">{item.title}</h3>
          <p className="text-gray-500">Brand: {item.brand}</p>
          <p className="text-gray-700">₹{item.price}</p>
        </div>
      </div>
      <div className="flex items-center space-x-4 w-full justify-between md:w-auto">
        <div className="flex items-center space-x-2">
          <button
            onClick={() => onDecrease(item.id)}
            className="p-2 bg-gray-200 rounded-full hover:bg-gray-300"
            aria-label={`Decrease quantity of ${item.title}`}
            data-testid="decrease-button"
          >
            <FaMinus />
          </button>
          <span className="text-gray-700" aria-live="polite">{item.quantity}</span>
          <button
            onClick={() => onIncrease(item.id)}
            className="p-2 bg-gray-200 rounded-full hover:bg-gray-300"
            aria-label={`Increase quantity of ${item.title}`}
            data-testid="increase-button"
          >
            <FaPlus />
          </button>
        </div>
        <p className="text-gray-700">
          ₹{(item.price * item.quantity).toFixed(2)}
        </p>
        <button
          onClick={() => onRemove(item.id)}
          className="text-red-500 hover:text-red-700"
          aria-label={`Remove ${item.title} from cart`}
          data-testid="remove-button"
        >
          <FaTrashAlt />
        </button>
      </div>
    </div>
  )
}

export default CartItem
