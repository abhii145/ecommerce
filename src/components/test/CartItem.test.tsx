import React from "react"
import { render, fireEvent } from "@testing-library/react"
import CartItem from "../CartItem"

const mockItem = {
  id: 1,
  title: "Sample Item",
  brand: "Sample Brand",
  price: 100,
  category: "Sample Category",
  quantity: 2,
  thumbnail: "sample-thumbnail.jpg",
  images: ["sample-image1.jpg", "sample-image2.jpg"],
}

describe("CartItem", () => {
  const onIncrease = jest.fn()
  const onDecrease = jest.fn()
  const onRemove = jest.fn()

  const renderComponent = () =>
    render(
      <CartItem
        item={mockItem}
        onIncrease={onIncrease}
        onDecrease={onDecrease}
        onRemove={onRemove}
      />
    )
  
   it("matches snapshot", () => {
     const { asFragment } = render(
       <CartItem
         item={mockItem}
         onIncrease={onIncrease}
         onDecrease={onDecrease}
         onRemove={onRemove}
       />
     )
     expect(asFragment()).toMatchSnapshot()
   })


  it("renders correctly", () => {
    const { getByText, getByAltText } = renderComponent()
    expect(getByText("Sample Item")).toBeInTheDocument()
    expect(getByText("Brand: Sample Brand")).toBeInTheDocument()
    expect(getByText("₹100")).toBeInTheDocument()
    expect(getByText("₹200.00")).toBeInTheDocument()
    expect(getByAltText("Sample Item")).toBeInTheDocument()
  })

  it("calls onIncrease when the increase button is clicked", () => {
    const { getByTestId } = renderComponent()
    fireEvent.click(getByTestId("increase-button"))
    expect(onIncrease).toHaveBeenCalledWith(mockItem.id)
  })

  it("calls onDecrease when the decrease button is clicked", () => {
    const { getByTestId } = renderComponent()
    fireEvent.click(getByTestId("decrease-button"))
    expect(onDecrease).toHaveBeenCalledWith(mockItem.id)
  })

  it("calls onRemove when the remove button is clicked", () => {
    const { getByTestId } = renderComponent()
    fireEvent.click(getByTestId("remove-button"))
    expect(onRemove).toHaveBeenCalledWith(mockItem.id)
  })

  it("displays the correct quantity", () => {
    const { getByText } = renderComponent()
    expect(getByText("2")).toBeInTheDocument()
  })

  it("displays the correct total price", () => {
    const { getByText } = renderComponent()
    expect(getByText("₹200.00")).toBeInTheDocument()
  })
})
