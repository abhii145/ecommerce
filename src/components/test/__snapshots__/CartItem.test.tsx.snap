// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col md:flex-row items-center justify-between border-b py-4"
  >
    <div
      class="flex items-center space-x-4 mb-4 md:mb-0"
    >
      <img
        alt="Sample Item"
        class="w-24 h-24 object-cover rounded-md"
        loading="lazy"
        src="sample-thumbnail.jpg"
      />
      <div>
        <h3
          class="text-lg font-semibold"
        >
          Sample Item
        </h3>
        <p
          class="text-gray-500"
        >
          Brand: Sample Brand
        </p>
        <p
          class="text-gray-700"
        >
          ₹100
        </p>
      </div>
    </div>
    <div
      class="flex items-center space-x-4 w-full justify-between md:w-auto"
    >
      <div
        class="flex items-center space-x-2"
      >
        <button
          aria-label="Decrease quantity of Sample Item"
          class="p-2 bg-gray-200 rounded-full hover:bg-gray-300"
          data-testid="decrease-button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
            />
          </svg>
        </button>
        <span
          aria-live="polite"
          class="text-gray-700"
        >
          2
        </span>
        <button
          aria-label="Increase quantity of Sample Item"
          class="p-2 bg-gray-200 rounded-full hover:bg-gray-300"
          data-testid="increase-button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
            />
          </svg>
        </button>
      </div>
      <p
        class="text-gray-700"
      >
        ₹200.00
      </p>
      <button
        aria-label="Remove Sample Item from cart"
        class="text-red-500 hover:text-red-700"
        data-testid="remove-button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
